File: zig/src/storage/backend.zig
const std = @import("std");

/// Result of a getObject or headObject call.
pub const ObjectData = struct {
    body: ?[]const u8,
    content_length: u64,
    content_type: []const u8,
    etag: []const u8,
    last_modified: []const u8,
};

/// Options for a putObject call.
pub const PutObjectOptions = struct {
    content_type: []const u8 = "application/octet-stream",
    content_length: u64 = 0,
};

/// Result of a putObject call.
pub const PutObjectResult = struct {
    etag: []const u8,
};

/// Describes a single completed part in a multipart upload.
pub const PartInfo = struct {
    part_number: u32,
    etag: []const u8,
};

/// Result of a putPart call.
pub const PutPartResult = struct {
    etag: []const u8,
};

/// Result of an assembleParts call.
pub const AssemblePartsResult = struct {
    etag: []const u8,
    total_size: u64 = 0,
};

/// StorageBackend interface using a vtable pattern.
///
/// Concrete implementations (LocalBackend, AwsGatewayBackend, etc.) populate
/// the vtable and store their own state behind the `ctx` opaque pointer.
pub const StorageBackend = struct {
    ctx: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        putObject: *const fn (ctx: *anyopaque, bucket: []const u8, key: []const u8, data: []const u8, opts: PutObjectOptions) anyerror!PutObjectResult,
        getObject: *const fn (ctx: *anyopaque, bucket: []const u8, key: []const u8) anyerror!ObjectData,
        deleteObject: *const fn (ctx: *anyopaque, bucket: []const u8, key: []const u8) anyerror!void,
        headObject: *const fn (ctx: *anyopaque, bucket: []const u8, key: []const u8) anyerror!ObjectData,
        copyObject: *const fn (ctx: *anyopaque, src_bucket: []const u8, src_key: []const u8, dst_bucket: []const u8, dst_key: []const u8) anyerror!PutObjectResult,
        putPart: *const fn (ctx: *anyopaque, bucket: []const u8, upload_id: []const u8, part_number: u32, data: []const u8) anyerror!PutPartResult,
        assembleParts: *const fn (ctx: *anyopaque, bucket: []const u8, key: []const u8, upload_id: []const u8, parts: []const PartInfo) anyerror!AssemblePartsResult,
        deleteParts: *const fn (ctx: *anyopaque, bucket: []const u8, upload_id: []const u8) anyerror!void,
        createBucket: *const fn (ctx: *anyopaque, bucket: []const u8) anyerror!void,
        deleteBucket: *const fn (ctx: *anyopaque, bucket: []const u8) anyerror!void,
    };

    // --- Convenience wrappers ---

    pub fn putObject(self: StorageBackend, bucket: []const u8, key: []const u8, data: []const u8, opts: PutObjectOptions) !PutObjectResult {
        return self.vtable.putObject(self.ctx, bucket, key, data, opts);
    }

    pub fn getObject(self: StorageBackend, bucket: []const u8, key: []const u8) !ObjectData {
        return self.vtable.getObject(self.ctx, bucket, key);
    }

    pub fn deleteObject(self: StorageBackend, bucket: []const u8, key: []const u8) !void {
        return self.vtable.deleteObject(self.ctx, bucket, key);
    }

    pub fn headObject(self: StorageBackend, bucket: []const u8, key: []const u8) !ObjectData {
        return self.vtable.headObject(self.ctx, bucket, key);
    }

    pub fn copyObject(self: StorageBackend, src_bucket: []const u8, src_key: []const u8, dst_bucket: []const u8, dst_key: []const u8) !PutObjectResult {
        return self.vtable.copyObject(self.ctx, src_bucket, src_key, dst_bucket, dst_key);
    }

    pub fn putPart(self: StorageBackend, bucket: []const u8, upload_id: []const u8, part_number: u32, data: []const u8) !PutPartResult {
        return self.vtable.putPart(self.ctx, bucket, upload_id, part_number, data);
    }

    pub fn assembleParts(self: StorageBackend, bucket: []const u8, key: []const u8, upload_id: []const u8, parts: []const PartInfo) !AssemblePartsResult {
        return self.vtable.assembleParts(self.ctx, bucket, key, upload_id, parts);
    }

    pub fn deleteParts(self: StorageBackend, bucket: []const u8, upload_id: []const u8) !void {
        return self.vtable.deleteParts(self.ctx, bucket, upload_id);
    }

    pub fn createBucket(self: StorageBackend, bucket: []const u8) !void {
        return self.vtable.createBucket(self.ctx, bucket);
    }

    pub fn deleteBucket(self: StorageBackend, bucket: []const u8) !void {
        return self.vtable.deleteBucket(self.ctx, bucket);
    }
};

File: rust/src/storage/backend.rs
//! Abstract storage backend trait.
//!
//! Every storage backend must implement [`StorageBackend`].  The trait
//! works in terms of opaque byte streams so callers do not need to know
//! the underlying medium.

use bytes::Bytes;
use std::future::Future;
use std::pin::Pin;

/// A stored object's data plus its content hash.
#[derive(Debug, Clone)]
pub struct StoredObject {
    /// Raw bytes of the object.
    pub data: Bytes,
    /// Hex-encoded content hash (typically SHA-256).
    pub content_hash: String,
}

/// Async object storage contract.
pub trait StorageBackend: Send + Sync + 'static {
    /// Write `data` to `storage_key`, returning the content hash.
    fn put(
        &self,
        storage_key: &str,
        data: Bytes,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<String>> + Send + '_>>;

    /// Read the full object at `storage_key`.
    fn get(
        &self,
        storage_key: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<StoredObject>> + Send + '_>>;

    /// Delete the object at `storage_key`.
    fn delete(
        &self,
        storage_key: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + '_>>;

    /// Check whether an object exists at `storage_key`.
    fn exists(
        &self,
        storage_key: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<bool>> + Send + '_>>;

    /// Copy an object from `src_key` in `bucket` to `dst_key` in `dst_bucket`,
    /// returning the ETag of the new object.
    fn copy_object(
        &self,
        bucket: &str,
        src_key: &str,
        dst_bucket: &str,
        dst_key: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<String>> + Send + '_>>;

    /// Write a single part for a multipart upload, returning the part ETag.
    fn put_part(
        &self,
        bucket: &str,
        upload_id: &str,
        part_number: u32,
        data: Bytes,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<String>> + Send + '_>>;

    /// Assemble previously uploaded parts into a final object, returning
    /// the ETag of the completed object.
    fn assemble_parts(
        &self,
        bucket: &str,
        key: &str,
        upload_id: &str,
        parts: &[(u32, String)],
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<String>> + Send + '_>>;

    /// Delete all part data associated with a multipart upload.
    fn delete_parts(
        &self,
        bucket: &str,
        upload_id: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + '_>>;

    /// Create a new storage bucket.
    fn create_bucket(
        &self,
        bucket: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + '_>>;

    /// Delete a storage bucket.
    fn delete_bucket(
        &self,
        bucket: &str,
    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + '_>>;
}
